src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+                                                          const ShellCommand& command)
+   {
+      boost::shared_ptr<ExecuteChunkOperation> pProcess =
+            boost::shared_ptr<ExecuteChunkOperation>(new ExecuteChunkOperation(
+                                                        docId,
+                                                        chunkId,
+                                                        command));
+      pProcess->registerProcess();
+      return pProcess;
+   }
+   
+private:
+   
+   ExecuteChunkOperation(const std::string& docId,
+                         const std::string& chunkId,
+                         const ShellCommand& command)
+      : isRunning_(false),
+        terminationRequested_(false),
+        docId_(docId),
+        chunkId_(chunkId),
+        command_(command)
+   {
+      using namespace core;
+      Error error = Success();
+      
+      // ensure staging directory
+      FilePath stagingPath = chunkOutputPath(
+               docId_,
+               chunkId_ + kStagingSuffix,
+               ContextExact);
+      
+      error = stagingPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = stagingPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // ensure regular directory
+      FilePath outputPath = chunkOutputPath(
+               docId_,
+               chunkId_,
+               ContextExact);
+      
+      error = outputPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = outputPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // clean old chunk output
+      error = cleanChunkOutput(docId_, chunkId_, true);
+      if (error)
+         LOG_ERROR(error);
+   }
+   
+public:
+   
+   ProcessCallbacks processCallbacks()
+   {
+      ProcessCallbacks callbacks;
+      
+      callbacks.onStarted = boost::bind(
+               &ExecuteChunkOperation::onStarted,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onContinue = boost::bind(
+               &ExecuteChunkOperation::onContinue,
+               shared_from_this(),
+               _1);
+      
+      callbacks.onStdout = boost::bind(
+               &ExecuteChunkOperation::onStdout,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onStderr = boost::bind(
+               &ExecuteChunkOperation::onStderr,
+               shared_from_this(),
+               _1,
+               _2);
+      
+      callbacks.onExit = boost::bind(
+               &ExecuteChunkOperation::onExit,
+               shared_from_this(),
+               _1);
+      
+      return callbacks;
+   }
+   
+private:
+   
+   enum OutputType { OUTPUT_STDOUT, OUTPUT_STDERR };
+   
+   void onStarted(ProcessOperations& operations)
+   {
+      isRunning_ = true;
```

Comment:
Depending on what isRunning_ is used for you may want to set this earlier (since there can be a delay before onStarted is called due to polling intervals). Perhaps you could have an `init()` method that is called from the static creation method which sets this value and also calls `registerProcess`